### PR TITLE
Use reverse_bits intrinsic when supported

### PR DESCRIPTION
**Pull request by thomwiggers**
This adds a new feature that allows to conditionally use the new
reverse_bits feature in Rust.

### CODE REVIEW

**Comment by ltratt** [comment]
Does this provide a noticeable speed-up?

**Comment by ltratt** [comment]
Thinking about this a bit more: assuming this does give a useful speed-up (which I imagine it does, but we need to see numbers to be sure), wouldn't it be easier to use this automatically if `nightly` is used to build `vob`? Put another way: if it works well, I can't imagine anyone not wanting to use this feature. This would also make it completely transparent to the user, and mean that we don't have any future headaches removing a Cargo feature which some people have come to rely on.

**Comment by thomwiggers** [comment]
> Does this provide a noticeable speed-up?
CI is running benchmarks when running builds on `nightly`; so you should be able to see the numbers there.

There is a slight speed-up on my system, but it should be more noticeable on e.g. ARMv8 where there is CPU support.

> wouldn't it be easier to use this automatically if nightly is used to build vob? Put another way: if it works well, I can't imagine anyone not wanting to use this feature. This would also make it completely transparent to the user, and mean that we don't have any future headaches removing a Cargo feature which some people have come to rely on.

Unfortunately, I don't know of a way to detect if a feature is available in 'normal' rust. There's ways to set a feature flag using a `build.rs`, but that seems a bit excessive. Also, once it stabilizes, you kind of still want to build with the feature flag turned on to opt in to the new behaviour - but detecting nightly would not be sufficient to do that anymore. 

See also https://internals.rust-lang.org/t/setting-cfg-nightly-on-nightly-by-default/1893

**Comment by ltratt** [comment]
What sort of speed-up did you see? And, assuming it's an improvement, why would we want to make this an opt-in feature? It feels like it should be something the user doesn't know about: it just makes their code go faster.

**Comment by thomwiggers** [comment]
4k ns to 3k or something.

It's not that I wouldn't want to, it's just that I can't figure out how to do it.

**Comment by ltratt** [comment]
I must admit, for the complexity involved, I'm still wondering if we should just wait for this to stabilise. Admittedly, I'm not sure *when* it will stabilise.

**Comment by ltratt** [comment]
In the interests of not letting things languish without a decision forever I think we should say that we won't merge this unless it can be detected automatically (with `build.sh` or some other trick). And if we don't want to do such a trick, it's probably better to close the PR. Sound like a plan?

**Comment by thomwiggers** [comment]
That makes sense. Let's mark it WIP for now.

**Comment by ltratt** [comment]
OK, I'll close this, but of course you can reopen it if/when automatic detection works.

**Comment by thomwiggers** [comment]
Ok, I had some procrastination from writing on my master's thesis to do, so I implemented detection of nightly on top of the feature flags that are available.

There's now a `nightly` cfg flag that is automatically set on nightly. Its use is to enable the `#![feature]` tags. Those will only ever work on nightly, anyway.

The `nightly` flag is separate from the `reverse_bits` *cfg flag*. That flag will be set if nightly is detected and the `no_reverse_bits` feature is not enabled, or if the `reverse_bits` *feature* is enabled. `build.rs` contains a sanity check that both aren't enabled at the same time.

The `reverse_bits` feature will *not* switch on the `#[feature]` flag, to allow switching it on in `stable`. Of course, that won't work for the time being, but it will allow people that use Vob to opt-in to the new behaviour whenever https://github.com/rust-lang/rust/issues/48763 is stabilized, without a Vob update. (Though we should obviously still keep track of that issue)

By the way, I recommend switching on the `cron` settings of Travis to check nightly builds daily.

File: build.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+fn main() {
+    assert!(
+        !(env::var("CARGO_FEATURE_REVERSE_BITS").is_ok()
+            && env::var("CARGO_FEATURE_NO_REVERSE_BITS").is_ok()),
+        "Can't both enable and disable REVERSE_BITS"
+    );
+
+    // Set features depending on channel
+    match version_meta().unwrap().channel {
```
Comment (ltratt) [inline build.rs]: Would this better be `if let Channel::Nightly = version_meta().unwrap().channel { ... } else { ... }`?

File: Cargo.toml [hunk 2]
```diff
@@ -20,3 +23,4 @@ rand = "0.5"
 [features]
 unsafe_internals = []
 reverse_bits = []
```
Comment (ltratt) [inline Cargo.toml]: I don't think we need `reverse_bits` and `no_reverse_bits`. Personally I'd delete both, and just silently use reverse bits; but if we really, really want an option I think `no_reverse_bits` is enough.

**Comment by ltratt** [comment]
Are we not able to query rustc directly for whether the `reverse_bits` features is available?

**Comment by thomwiggers** [comment]
> Are we not able to query rustc directly for whether the reverse_bits features is available?

AFAIK, no. You could perhaps trail-compile something, but that seems way too much. That's why I included the `reverse_bits` feature.

**Comment by ltratt** [comment]
@vext01 Do you know if you can query rustc for a feature directly?

@thomwiggers If not, I agree your approach is reasonable.

**Comment by vext01** [comment]
You mean like:

```
#[cfg(feature = "foo")]
```
?

https://doc.rust-lang.org/1.26.0/book/first-edition/conditional-compilation.html

**Comment by ltratt** [comment]
Not exactly. We'd like to detect when this feature https://doc.rust-lang.org/std/primitive.isize.html#method.reverse_bits is available or not (at the moment we assume "if we're on nightly, it's available", which might be a bit fragile).

**Comment by vext01** [comment]
What do you want to do if it's not available?

**Comment by ltratt** [comment]
If it's not available, we have a fallback (slower) route that doesn't involve using the feature.

**Comment by vext01** [comment]
OK, so it would look like this I think:
```
#[cfg(feature = "reverse_bits")] {
    // fast code
}
#[cfg(not(feature = "reverse_bits"))] {
    // slow code
}
```

**Comment by thomwiggers** [comment]
That only works for `Cargo.toml`-defined features.

**Comment by vext01** [comment]
Ugh. That being the case, no idea then. Sorry.

**Comment by vext01** [comment]
(I reckon that is a bug in rustc, why should compiler features be special?)

**Comment by vext01** [comment]
The only other thing you can do is make a feature check in a build.rs, but that's horrible too.

**Comment by ltratt** [comment]
OK, then I think Thom's suggestion is as good as it gets. We can live with it. There are still a couple of other comments for @thomwiggers though :)

File: .travis.yml [hunk 3]
```diff
@@ -1,13 +1,27 @@
 language: rust
 
+rust:
+  - stable
+  - beta
```
Comment (ltratt) [inline .travis.yml]: I must admit, I'm not very sure it's useful for us to test on `beta`. I know it doesn't *hurt*, but I feel a bit bad for using electricity for something that's probably never going to show an error.

Comment (thomwiggers) [reply .travis.yml]: Seems like a good way to eliminate a lot of builds as well

Comment (thomwiggers) [reply .travis.yml]: done.

File: Cargo.toml (hunk 2 above)
Comment (thomwiggers) [reply Cargo.toml]: deleted the features.

File: build.rs (hunk 1 above)
Comment (thomwiggers) [reply build.rs]: done, especially since we no longer handle the feature settings.

File: Cargo.toml [hunk 4]
```diff
@@ -19,3 +22,5 @@ rand = "0.5"
 
 [features]
 unsafe_internals = []
+reverse_bits = []
```
Comment (ltratt) [inline Cargo.toml]: Are these two lines supposed to be deleted?

Comment (thomwiggers) [reply Cargo.toml]: hmm somehow I missed them

Comment (ltratt) [reply Cargo.toml]: I thought you might be testing me :p

**Comment by ltratt** [comment]
OK, I think this is ready to be squashed.

**Comment by thomwiggers** [comment]
It's been squashed by the way

File: build.rs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+extern crate rustc_version;
+
+use rustc_version::{version_meta, Channel};
+use std::env;
+
+fn main() {
+    assert!(
+        !(env::var("CARGO_FEATURE_REVERSE_BITS").is_ok()
```
Comment (thomwiggers) [inline build.rs]: oh wait, this needs to go.

Comment (thomwiggers) [reply build.rs]: done.

**Comment by ltratt** [comment]
Thanks!

**Pull request merged**